src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/LatencyAndInflightCountLoadBalancingPolicy.java

```diff
@@ -0,0 +1,289 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.loadbalancing;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+
+import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
+import com.datastax.oss.driver.api.core.context.DriverContext;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.session.Request;
+import com.datastax.oss.driver.api.core.session.Session;
+import com.datastax.oss.driver.internal.core.pool.ChannelPool;
+import com.datastax.oss.driver.internal.core.session.DefaultSession;
+import com.datastax.oss.driver.internal.core.util.ArrayUtils;
+import com.datastax.oss.driver.internal.core.util.collection.QueryPlan;
+import com.datastax.oss.driver.internal.core.util.collection.SimpleQueryPlan;
+import com.datastax.oss.driver.shaded.guava.common.cache.CacheBuilder;
+import com.datastax.oss.driver.shaded.guava.common.cache.CacheLoader;
+import com.datastax.oss.driver.shaded.guava.common.cache.LoadingCache;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LatencyAndInflightCountLoadBalancingPolicy extends DefaultLoadBalancingPolicy {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(LatencyAndInflightCountLoadBalancingPolicy.class);
+
+  protected final LoadingCache<Node, NodeLatencyTracker> latencies;
+
+  private final long RETRY_PERIOD = TimeUnit.SECONDS.toNanos(2);
+
+  private final long THRESHOLD_TO_ACCOUNT = 100;
+
+  // Just a copy of 4.18.0 java driver NEWLY_UP_INTERVAL_NANOS
+  private static final long NEWLY_UP_INTERVAL = MINUTES.toNanos(1);
+
+  public LatencyAndInflightCountLoadBalancingPolicy(
+      @NonNull DriverContext context, @NonNull String profileName) {
+    super(context, profileName);
+    CacheLoader<Node, NodeLatencyTracker> cacheLoader =
+        new CacheLoader<Node, NodeLatencyTracker>() {
+          @Override
+          public NodeLatencyTracker load(@NonNull Node node) {
+            return new NodeLatencyTracker(THRESHOLD_TO_ACCOUNT);
+          }
+        };
+    latencies = CacheBuilder.newBuilder().weakKeys().build(cacheLoader);
+  }
+
+  @NonNull
+  @Override
+  public Queue<Node> newQueryPlan(@Nullable Request request, @Nullable Session session) {
+    // Take a snapshot since the set is concurrent:
```

Comment:
Is there a reason we don't kick this up to super.newQueryPlan() if slow replica avoidance isn't on?

```
    if (!avoidSlowReplicas) {
      return super.newQueryPlan(request, session);
    }
```